File: apps/api/src/lib/crypto.ts
import crypto from 'crypto';
import { config } from './config.js';

// AES-256-GCM encryption parameters
const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 12; // 96 bits for GCM
const AUTH_TAG_LENGTH = 16;

/**
 * Get the encryption key for OAuth tokens (hex string to buffer)
 */
function getOAuthEncryptionKey(): Buffer {
  const keyHex = config.oauthEncryptionKey;
  if (keyHex.length !== 64) {
    throw new Error('OAUTH_ENCRYPTION_KEY must be 64 hex characters (32 bytes)');
  }
  return Buffer.from(keyHex, 'hex');
}

/**
 * Encrypt a string using AES-256-GCM
 * Returns a string in format: iv:authTag:ciphertext (all base64)
 */
export function encryptOAuthToken(plaintext: string): string {
  const key = getOAuthEncryptionKey();
  const iv = crypto.randomBytes(IV_LENGTH);

  const cipher = crypto.createCipheriv(ALGORITHM, key, iv, {
    authTagLength: AUTH_TAG_LENGTH,
  });

  const encrypted = Buffer.concat([
    cipher.update(plaintext, 'utf8'),
    cipher.final(),
  ]);

  const authTag = cipher.getAuthTag();

  // Combine IV + auth tag + ciphertext, all base64 encoded
  return `${iv.toString('base64')}:${authTag.toString('base64')}:${encrypted.toString('base64')}`;
}

/**
 * Decrypt an OAuth token encrypted with encryptOAuthToken()
 */
export function decryptOAuthToken(encryptedString: string): string {
  const key = getOAuthEncryptionKey();
  const parts = encryptedString.split(':');

  if (parts.length !== 3) {
    throw new Error('Invalid encrypted token format');
  }

  const [ivBase64, authTagBase64, ciphertextBase64] = parts;
  const iv = Buffer.from(ivBase64, 'base64');
  const authTag = Buffer.from(authTagBase64, 'base64');
  const ciphertext = Buffer.from(ciphertextBase64, 'base64');

  const decipher = crypto.createDecipheriv(ALGORITHM, key, iv, {
    authTagLength: AUTH_TAG_LENGTH,
  });
  decipher.setAuthTag(authTag);

  const decrypted = Buffer.concat([
    decipher.update(ciphertext),
    decipher.final(),
  ]);

  return decrypted.toString('utf8');
}

/**
 * Encrypt an OAuth token if it exists, otherwise return null/undefined
 */
export function encryptOAuthTokenOptional(token: string | null | undefined): string | null {
  if (!token) return null;
  return encryptOAuthToken(token);
}

/**
 * Decrypt an OAuth token if it exists and is in encrypted format.
 * Throws on decryption failure instead of silently returning raw value,
 * to prevent tampered or corrupted data from being used.
 */
export function decryptOAuthTokenOptional(token: string | null | undefined): string | null {
  if (!token) return null;

  // Check if token is in encrypted format (iv:authTag:ciphertext with base64)
  const parts = token.split(':');
  if (parts.length === 3 && parts.every(p => /^[A-Za-z0-9+/=]+$/.test(p))) {
    return decryptOAuthToken(token);
  }

  // Not in encrypted format — legacy unencrypted token (backward compatibility)
  return token;
}

// ============================================
// AWS External ID Encryption
// Uses the same encryption key as OAuth tokens for simplicity
// External IDs are sensitive as they're used for cross-account AWS access
// ============================================

/**
 * Encrypt an AWS external ID using AES-256-GCM
 * Returns a string in format: iv:authTag:ciphertext (all base64)
 */
export function encryptExternalId(plaintext: string): string {
  const key = getOAuthEncryptionKey();
  const iv = crypto.randomBytes(IV_LENGTH);

  const cipher = crypto.createCipheriv(ALGORITHM, key, iv, {
    authTagLength: AUTH_TAG_LENGTH,
  });

  const encrypted = Buffer.concat([
    cipher.update(plaintext, 'utf8'),
    cipher.final(),
  ]);

  const authTag = cipher.getAuthTag();

  return `${iv.toString('base64')}:${authTag.toString('base64')}:${encrypted.toString('base64')}`;
}

/**
 * Decrypt an AWS external ID encrypted with encryptExternalId()
 */
export function decryptExternalId(encryptedString: string): string {
  const key = getOAuthEncryptionKey();
  const parts = encryptedString.split(':');

  if (parts.length !== 3) {
    throw new Error('Invalid encrypted external ID format');
  }

  const [ivBase64, authTagBase64, ciphertextBase64] = parts;
  const iv = Buffer.from(ivBase64, 'base64');
  const authTag = Buffer.from(authTagBase64, 'base64');
  const ciphertext = Buffer.from(ciphertextBase64, 'base64');

  const decipher = crypto.createDecipheriv(ALGORITHM, key, iv, {
    authTagLength: AUTH_TAG_LENGTH,
  });
  decipher.setAuthTag(authTag);

  const decrypted = Buffer.concat([
    decipher.update(ciphertext),
    decipher.final(),
  ]);

  return decrypted.toString('utf8');
}

/**
 * Encrypt an external ID if it exists, otherwise return null
 */
export function encryptExternalIdOptional(id: string | null | undefined): string | null {
  if (!id) return null;
  return encryptExternalId(id);
}

/**
 * Decrypt an external ID if it exists and is in encrypted format.
 * Throws on decryption failure instead of silently returning raw value,
 * to prevent tampered or corrupted data from being used.
 */
export function decryptExternalIdOptional(id: string | null | undefined): string | null {
  if (!id) return null;

  // Check if ID is in encrypted format (iv:authTag:ciphertext with base64)
  const parts = id.split(':');
  if (parts.length === 3 && parts.every(p => /^[A-Za-z0-9+/=]+$/.test(p))) {
    return decryptExternalId(id);
  }

  // Not in encrypted format — legacy unencrypted ID (backward compatibility)
  return id;
}

File: workers/internal/crypto/crypto.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"strings"
)

// Decryptor handles AES-256-GCM decryption of encrypted values.
type Decryptor struct {
	key []byte
}

// NewDecryptor creates a new Decryptor with the given hex-encoded key.
// The key must be 64 hex characters (32 bytes) for AES-256.
func NewDecryptor(hexKey string) (*Decryptor, error) {
	if hexKey == "" {
		return nil, errors.New("encryption key is required")
	}
	if len(hexKey) != 64 {
		return nil, errors.New("encryption key must be 64 hex characters (32 bytes)")
	}

	key, err := hex.DecodeString(hexKey)
	if err != nil {
		return nil, errors.New("invalid hex encryption key")
	}

	return &Decryptor{key: key}, nil
}

// DecryptExternalID decrypts an external ID encrypted with AES-256-GCM.
// The encrypted format is: iv:authTag:ciphertext (all base64 encoded).
// If the value doesn't look encrypted (no colons or wrong format), returns it as-is.
func (d *Decryptor) DecryptExternalID(encrypted string) (string, error) {
	if encrypted == "" {
		return "", nil
	}

	// Check if it looks like encrypted format (iv:authTag:ciphertext)
	parts := strings.Split(encrypted, ":")
	if len(parts) != 3 {
		// Not encrypted, return as-is (backward compatibility)
		return encrypted, nil
	}

	// Validate that all parts look like base64
	for _, part := range parts {
		if !isBase64(part) {
			// Not encrypted, return as-is
			return encrypted, nil
		}
	}

	// Decode base64 parts
	iv, err := base64.StdEncoding.DecodeString(parts[0])
	if err != nil {
		return encrypted, nil // Not valid base64, return as-is
	}

	authTag, err := base64.StdEncoding.DecodeString(parts[1])
	if err != nil {
		return encrypted, nil
	}

	ciphertext, err := base64.StdEncoding.DecodeString(parts[2])
	if err != nil {
		return encrypted, nil
	}

	// Create AES cipher
	block, err := aes.NewCipher(d.key)
	if err != nil {
		return "", err
	}

	// Create GCM mode
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	// In GCM, the auth tag is appended to ciphertext for decryption
	ciphertextWithTag := append(ciphertext, authTag...)

	// Decrypt
	plaintext, err := gcm.Open(nil, iv, ciphertextWithTag, nil)
	if err != nil {
		// Decryption failed, might be legacy unencrypted value
		return encrypted, nil
	}

	return string(plaintext), nil
}

// isBase64 checks if a string looks like valid base64.
func isBase64(s string) bool {
	if s == "" {
		return false
	}
	for _, c := range s {
		if !((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '+' || c == '/' || c == '=') {
			return false
		}
	}
	return true
}
